velox/connectors/hive/PartitionIdGenerator.cpp

```diff
@@ -21,44 +21,135 @@ namespace facebook::velox::connector::hive {
+  // partition. VectorHashers keep track of the number of unique values, hence,
+  // we can find out if there is only one unique value for each partition key.
+  for (auto i = 0; i < numRows; ++i) {
+    auto valueId = result[i];
+    auto it = partitionIds_.find(valueId);
+    if (it != partitionIds_.end()) {
+      result[i] = it->second;
+    } else {
+      uint64_t nextPartitionId = partitionIds_.size();
+      VELOX_USER_CHECK_LT(
+          nextPartitionId,
+          maxPartitions_,
+          "Exceeded limit of distinct partitions.");
+
+      partitionIds_.emplace(valueId, nextPartitionId);
+      savePartitionKeys(nextPartitionId, input, i);
+
+      result[i] = nextPartitionId;
+    }
+  }
+
+  recentMaxId_ = *std::max_element(result.begin(), result.end());
+}
+
+void PartitionIdGenerator::computeValueIds(
+    const RowVectorPtr& input,
+    raw_vector<uint64_t>& valueIds) {
   allRows_.resize(input->size());
   allRows_.setAll();
 
-  auto partitionVector = input->childAt(hasher_->channel())->loadedVector();
-  hasher_->decode(*partitionVector, allRows_);
+  bool rehash = false;
+  for (auto& hasher : hashers_) {
+    auto partitionVector = input->childAt(hasher->channel())->loadedVector();
+    hasher->decode(*partitionVector, allRows_);
+    if (!hasher->computeValueIds(allRows_, valueIds)) {
+      rehash = true;
+    }
+  }
+
+  if (rehash) {
+    uint64_t multiplier = 1;
+    for (auto& hasher : hashers_) {
+      multiplier = hasher->enableValueIds(multiplier, 50);
+
+      VELOX_CHECK_NE(
+          multiplier,
+          exec::VectorHasher::kRangeTooLarge,
+          "Number of requested IDs is out of range.");
+    }
 
-  if (!hasher_->computeValueIds(allRows_, result)) {
-    uint64_t range = hasher_->enableValueIds(1, kHasherReservePct);
-    VELOX_CHECK_NE(
-        range,
-        exec::VectorHasher::kRangeTooLarge,
-        "Number of requested IDs is out of range.");
+    for (auto& hasher : hashers_) {
+      bool ok = hasher->computeValueIds(allRows_, valueIds);
+      VELOX_CHECK(ok);
+    }
 
-    VELOX_CHECK(
-        hasher_->computeValueIds(allRows_, result),
-        "Cannot assign new value IDs.");
+    updateValueToPartitionIdMapping();
   }
+}
 
-  recentMaxId_ = *std::max_element(result.begin(), result.end());
-  maxId_ = std::max(maxId_, recentMaxId_);
+void PartitionIdGenerator::updateValueToPartitionIdMapping() {
+  if (partitionIds_.empty()) {
+    return;
+  }
+
+  auto numPartitions = partitionIds_.size();
 
-  VELOX_USER_CHECK_LE(
-      maxId_, maxPartitions_, "Exceeded limit of distinct partitions.");
+  partitionIds_.clear();
+
+  raw_vector<uint64_t> newValueIds(numPartitions);
+  SelectivityVector rows(numPartitions);
+  for (auto i = 0; i < hashers_.size(); ++i) {
+    auto& hasher = hashers_[i];
+    hasher->decode(*partitionKeys_->childAt(hasher->channel()), rows);
+    bool ok = hasher->computeValueIds(rows, newValueIds);
+    VELOX_CHECK(ok);
```

Comment:
Same here. We can do AND on ok. And do VELOX_CHECK on ok only once outside the loop.